Review Posts By Vote

Raemon opened the pull request:
This creates a page where users can view the posts that they strong upvoted in 2020, for the 2020 LW Review.

Right now this is focused on the Review. It may later make sense to make a more flexible page that can show other types of votes.

jimrandomh commented on packages/lesswrong/lib/collections/posts/views.ts:
```diff
@@ -1218,3 +1219,23 @@ Posts.addView("stickied", (terms: PostsViewTerms) => (
+  }
+})
+ensureIndex(Posts,
+  augmentForDefaultView({ postIds:1 }),
```
This isn't a field on `Posts`. The index should be covered by the `_id` index anyways.

jimrandomh commented on packages/lesswrong/components/posts/PostsByVote.tsx:
```diff
@@ -0,0 +1,59 @@
+    if (!votes) return null
+    const postIds = votes.map(vote=>vote.documentId)
+
+    const { results: posts } = useMulti({
```
Lint errors are real (due to the `currentUser` null check being before hooks). Either split this into two components the inner of which takes a nonnull user, or move the check down and put a `skip` on the query hooks.

jimrandomh commented on packages/lesswrong/lib/collections/posts/views.ts:
```diff
@@ -1218,3 +1219,23 @@ Posts.addView("stickied", (terms: PostsViewTerms) => (
+  return {
+    selector: {
+      _id: {$in: terms.postIds},
+      userId: {$ne: terms.userId},
```
This is a security vulnerability that enables deanonymizing all votes through the API

Raemon replied:
Wait how? (the pieces here don't actually intrinsically have anything to do with votes. Am I missing a larger interaction?)

jimrandomh replied:
Oops, this one isn't actually a vuln (I was thinking this one was also on the votes collection like the other one).

jimrandomh commented on packages/lesswrong/lib/collections/votes/fragments.ts:
```diff
@@ -24,3 +24,17 @@ registerFragment(`
+    cancelled
+    documentId
+    votedAt
+    isUnvote
```
Code smell here: What are `cancelled` and `isUnvote` doing in a fragment? Those fields are a sort of soft-delete mechanism, so the only way they can get exposed on the client is if the client is getting cancelled votes.

Raemon replied:
Cancelled was necessary to filter out cancelled votes. (I specifically ran into that problem while testing it)

I think I initially assumed I'd also have to filter out isUnvote, but it looks like I didn't get around to doing that, and I'm not 100% sure what it even means actually.

jimrandomh replied:
Filtering out cancelled votes should be happening server-side in the view, before the fragment becomes relevant. (Which I think is already what's happening.)

Raemon replied:
It wasn't happening, at least the way I was fetching the votes. (I've updated that this may require an overall different approach than I was doing)

It looks like currently most of the way deal with votes is through "currentUserVote" on makeVoteable, which is where the "cancelled-filter" lives.

jimrandomh commented on packages/lesswrong/lib/collections/votes/views.ts:
```diff
@@ -29,3 +35,20 @@ Votes.addView("tagVotes", function () {
+  return {
+    selector: {
+      collectionName: collectionName,
+      userId: userId,
```
This is a security vulnerability that deanonymizes votes

Raemon replied:
Does this tank the entire concept? 

I'm not familiar enough with this part of the architecture to have a clear sense of the right way to do this.

I see we have a "currentUserVote" applied via makeVoteable. If I want to get all my votes from a given time period (without knowing in advance which posts are connected to those votes), what's the best way to do that?

jimrandomh commented:
This PR contains two security vulnerabilities that deanonymize votes, both arising from the fact that views are being passed the current user's ID, but fail to enforce that it's actually the current user's ID rather than someone else's ID.

View functions actually get more arguments than just the `terms`, including a `ResolverContext`, so this can actually be fixed pretty easily without needing to rewire the view system.

Raemon commented:
@jimrandomh it looks like this PR was probably going about things the wrong way. I'd be interested in at some point chatting about what the right way to fetch all of a user's votes from a given year is.

jimrandomh commented:
It's actually only a small change! The `userPostVotes` view should take three arguments instead of one (all view functions do, it's just that the second and third are usually ignored); the third argument is a `ResolverContext`, which has the `userId` of the logged-in user on it. If you use that `userId` instead of getting one from the terms object, it's safe.

Raemon commented:
@jimrandomh I made some updates, partly addressing your concerns, partly reworking a bit so that it showed two different vote-lists (one for weak upvotes, one for strong). I realized that I'm actually an outlier in how much I use strong upvotes, and people vary in how strong a signal a weak upvote is.

Raemon commented:
I've created one (hopefully) final update that makes the url more flexible, taking in a "year" parameter instead of hardcoding "2020" into the codebase. Will allow users to look over their past voting history if they like, as well as make it easier to re-use next year.

@jimrandomh or @darkruby501 , take a look when you get a chance.

Raemon commented:
I'm merging this because it got at least one round of feedback, which I believe I have addressed. @jimrandomh should still doublecheck that there aren't any other unforeseen voting vulnerabilities introduced and that my decision to include isUnvote and cancelled in the fragment were correct. If there's an issue we can fix it in another PR.
